File: app/frontend/src/core/services/converters/FormatConverter.ts
import { EventEmitter } from 'events';
import { Readable, Writable } from 'stream';
import { ServiceResponse, ServiceError } from '../base/types';
import { StreamProcessingResult as MediaStreamProcessingResult } from '../types/MediaProcessing';

export { ServiceResponse };

export interface ConversionOptions {
  targetFormat: string;
  format?: string;  // For backward compatibility
  quality?: number;
  compression?: boolean;
  // Video-specific options
  fps?: number;
  width?: number;
  height?: number;
  videoCodec?: string;
  videoBitrate?: string;
  // Audio-specific options
  audioCodec?: string;
  audioBitrate?: string;
  audioChannels?: number;
  sampleRate?: number;
  // Stream processing options
  chunkSize?: number;
  codec?: string;
  preset?: string;
  bitrate?: string | number;
  metadata?: Record<string, string>;
  stripMetadata?: boolean;
  preserveMetadata?: boolean;
  tempDir?: string;
  stripExif?: boolean;
  compressionLevel?: number;
  memoryLimit?: number;
  preserveAspectRatio?: boolean;
}

export interface StreamOptions {
  chunkSize?: number;
  memoryLimit?: number;
  preserveMetadata?: boolean;
  quality?: number;
}

export interface ConversionProgress {
  bytesProcessed: number;
  totalBytes: number;
  percent: number;
  stage: string;
  timeElapsed?: number;
  timeRemaining?: number;
  currentFile?: string;
  memoryUsage: number;
}

export interface ConversionMetadata {
  format: string;
  size: number;
  width?: number;
  height?: number;
  duration?: number;
  bitrate?: string;
  codec?: string;
  fps?: string;
  audioCodec?: string;
  audioChannels?: number;
  audioSampleRate?: number;
  createdAt: Date;
  [key: string]: unknown;
}

export interface ConversionResult {
  data: Buffer;
  format: string;
  size: number;
  duration?: number;
  width?: number;
  height?: number;
  fps?: number;
  bitrate?: string;
  codec?: string;
}

export interface StreamProcessingResult extends MediaStreamProcessingResult {
  format: string;
  size: number;
}

export interface StreamConversionResult extends StreamProcessingResult {
  stream: Readable;
  format: string;
  size: number;
}

export interface ConversionStats {
  inputSize: number;
  outputSize: number;
  conversionTime: number;
  successCount: number;
  failureCount: number;
  totalConversions: number;
  peakMemoryUsage: number;
}

export type ConversionEventHandler<T> = (data: T) => void;

export interface FormatConverter {
  convert(input: Buffer | string, options: ConversionOptions): Promise<ServiceResponse<ConversionResult>>;
  convertStream(
    inputStream: Readable,
    outputStream: Writable,
    options: ConversionOptions
  ): Promise<ServiceResponse<StreamProcessingResult>>;
  getSupportedFormats(): string[];
  getSupportedInputFormats(): string[];
  getSupportedOutputFormats(): string[];
  canConvertTo(format: string): boolean;
  cleanup(): Promise<void>;
}

export abstract class BaseFormatConverter extends EventEmitter implements FormatConverter {
  abstract getSupportedInputFormats(): string[];
  abstract getSupportedOutputFormats(): string[];
  abstract convert(input: Buffer | string, options: ConversionOptions): Promise<ServiceResponse<ConversionResult>>;
  abstract canConvertTo(format: string): boolean;
  
  async cleanup(): Promise<void> {
    // Default implementation - can be overridden by specific converters
  }

  getSupportedFormats(): string[] {
    return [...this.getSupportedInputFormats(), ...this.getSupportedOutputFormats()];
  }

  abstract convertStream(
    inputStream: Readable,
    outputStream: Writable,
    options: ConversionOptions
  ): Promise<ServiceResponse<StreamProcessingResult>>;
}

/**
 * Type-safe event interface for format converters
 */
export interface FormatConverterEvents {
  progress: ConversionProgress;
  error: ServiceError;
  complete: ConversionResult | StreamProcessingResult;
} 

File: backend/app/04_very_high_complexity/src/core/services/converters/FormatConverter.py
from typing import Any, Dict, List, Union



{ ServiceResponse }
class ConversionOptions:
    targetFormat: str
    format?: str
    quality?: float
    compression?: bool
    fps?: float
    width?: float
    height?: float
    videoCodec?: str
    videoBitrate?: str
    audioCodec?: str
    audioBitrate?: str
    audioChannels?: float
    sampleRate?: float
    chunkSize?: float
    codec?: str
    preset?: str
    bitrate?: Union[str, float]
    metadata?: Dict[str, str>
    stripMetadata?: bool
    preserveMetadata?: bool
    tempDir?: str
    stripExif?: bool
    compressionLevel?: float
    memoryLimit?: float
    preserveAspectRatio?: bool
class StreamOptions:
    chunkSize?: float
    memoryLimit?: float
    preserveMetadata?: bool
    quality?: float
class ConversionProgress:
    bytesProcessed: float
    totalBytes: float
    percent: float
    stage: str
    timeElapsed?: float
    timeRemaining?: float
    currentFile?: str
    memoryUsage: float
class ConversionMetadata:
    format: str
    size: float
    width?: float
    height?: float
    duration?: float
    bitrate?: str
    codec?: str
    fps?: str
    audioCodec?: str
    audioChannels?: float
    audioSampleRate?: float
    createdAt: Date
    [key: str]: unknown
class ConversionResult:
    data: Buffer
    format: str
    size: float
    duration?: float
    width?: float
    height?: float
    fps?: float
    bitrate?: str
    codec?: str
class StreamProcessingResult:
    format: str
    size: float
class StreamConversionResult:
    stream: Readable
    format: str
    size: float
class ConversionStats:
    inputSize: float
    outputSize: float
    conversionTime: float
    successCount: float
    failureCount: float
    totalConversions: float
    peakMemoryUsage: float
type ConversionEventHandler<T> = (data: T) => void
class FormatConverter:
    convert(input: Union[Buffer, str, options: ConversionOptions): Awaitable[ServiceResponse<ConversionResult>>]
    convertStream(
    inputStream: Readable,
    outputStream: Writable,
    options: \'ConversionOptions\'
  ): Awaitable[ServiceResponse<StreamProcessingResult>>
    getSupportedFormats(): List[str]
    getSupportedInputFormats(): List[str]
    getSupportedOutputFormats(): List[str]
    canConvertTo(format: str): bool
    cleanup(): Awaitable[None>
abstract class BaseFormatConverter extends EventEmitter implements FormatConverter {
  abstract getSupportedInputFormats(): string[]
  abstract getSupportedOutputFormats(): string[]
  abstract convert(input: Buffer | string, options: ConversionOptions): Promise<ServiceResponse<ConversionResult>>
  abstract canConvertTo(format: str): bool
  async cleanup(): Promise<void> {
  }
  getSupportedFormats(): string[] {
    return [...this.getSupportedInputFormats(), ...this.getSupportedOutputFormats()]
  }
  abstract convertStream(
    inputStream: Readable,
    outputStream: Writable,
    options: \'ConversionOptions\'
  ): Promise<ServiceResponse<StreamProcessingResult>>
}
/**
 * Type-safe event interface for format converters
 */
class FormatConverterEvents:
    progress: \'ConversionProgress\'
    error: ServiceError
    complete: Union[ConversionResult, StreamProcessingResult] 